sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/BlobDecryptionPolicy.java

```diff
@@ -304,51 +352,121 @@ Flux<ByteBuffer> decryptBlob(String encryptedDataString, Flux<ByteBuffer> encryp
-            }
-            return null;
+    private Flux<ByteBuffer> decryptV2(Flux<ByteBuffer> encryptedFlux, EncryptionData encryptionData,
+        byte[] contentEncryptionKey) {
+        // Buffer an exact region with the nonce and tag
+        final int gcmEncryptionRegionLength = encryptionData.getEncryptedRegionInfo().getDataLength();
+        final int nonceLength = encryptionData.getEncryptedRegionInfo().getNonceLength();
+        BufferStagingArea stagingArea =
+            new BufferStagingArea(gcmEncryptionRegionLength + TAG_LENGTH + nonceLength,
+                gcmEncryptionRegionLength + TAG_LENGTH + nonceLength);
+
+        return UploadUtils.chunkSource(encryptedFlux,
+                new com.azure.storage.common.ParallelTransferOptions()
+                    .setBlockSizeLong((long) gcmEncryptionRegionLength
+                        + TAG_LENGTH + nonceLength))
+            .flatMapSequential(stagingArea::write)
+            .concatWith(Flux.defer(stagingArea::flush))
+            .flatMapSequential(aggregator -> {
+                // Get the IV out of the beginning of the aggregator
+                byte[] gmcIv = aggregator.getFirstNBytes(nonceLength);
+
+                Cipher gmcCipher;
+                try {
+                    gmcCipher = getCipher(contentEncryptionKey, encryptionData, gmcIv, false);
+                } catch (InvalidKeyException e) {
+                    return Mono.error(LOGGER.logExceptionAsError(Exceptions.propagate(e)));
+                }
+
+                ByteBuffer decryptedRegion = ByteBuffer.allocate(gcmEncryptionRegionLength);
+                return aggregator.asFlux()
+                    .map(buffer -> {
+                        // Write into the preallocated buffer and always return this buffer.
+                        try {
+                            gmcCipher.update(buffer, decryptedRegion);
+                        } catch (ShortBufferException e) {
+                            return Mono.error(LOGGER.logExceptionAsError(Exceptions.propagate(e)));
+                        }
+                        return decryptedRegion;
+                    }).then(Mono.fromCallable(() -> {
+                        // We have already written all the data to the cipher. Passing in a final
+                        // empty buffer allows us to force completion and return the filled buffer.
+                        gmcCipher.doFinal(ByteBuffer.allocate(0), decryptedRegion);
+                        decryptedRegion.flip();
+                        return decryptedRegion;
+                    })).flux();
+            });
+    }
+
+    private Flux<ByteBuffer> decryptV1(Flux<ByteBuffer> encryptedFlux, EncryptedBlobRange encryptedBlobRange,
+        boolean padding, EncryptionData encryptionData, String requestUri,
+        AtomicLong totalInputBytes, byte[] contentEncryptionKey) {
+        LOGGER.warning("Downloaded data found to be encrypted with v1 encryption, "
```

Comment:
This warning message is interesting to me, I think it's good to have but is it something that should happen every time a download is performed or only once during the lifetime of the application? An example of this is the JVM will log a warning/error message once during the lifetime of an application if a deprecated for removal JDK API is used. And maybe this is something we can do the same, or have the first log be a warning and subsequent logs be info/debug so that we don't spam a customer's logging system

Reply:
Interesting. I do think it'd be helpful to log somewhere any time we download an old encryption version since we log the uri, so it should be helpful in knowing what to upgrade, though that could maybe be moved to INFO or something. I'm not sure how we would manage to just log once, though, as we can't know until the blob is downloaded which version it is using, and I'm not sure how to track the state of whether we've already logged.